swiftsimio/accelerated.py

```diff
@@ -168,3 +180,148 @@ def index_dataset(handle: Dataset, mask_array: np.array) -> np.array:
     ranges = ranges_from_array(mask_array)
 
     return read_ranges_from_file(handle, ranges, output_size, output_type)
+
+
+################################ ALEXEI: playing around with better read_ranges_from_file implementation #################################
+
+def get_chunk_ranges(ranges, chunk_size, array_length):
+    chunk_ranges = []
+    for bound in ranges:
+        lower = int(np.floor(bound[0]/chunk_size))*chunk_size
+        upper = min(int(np.ceil(bound[1]/chunk_size))*chunk_size, array_length)
+
+        if len(chunk_ranges) > 0:
+            assert(lower >= chunk_ranges[-1][0])
+            assert(upper >= chunk_ranges[-1][1])
+            if lower > chunk_ranges[-1][1]:
+                chunk_ranges.append([lower, upper])
+            elif lower <= chunk_ranges[-1][1]:
+                chunk_ranges[-1][1] = upper
+            else:
+                raise RuntimeError("computing chunk ranges has gone horribly wrong")
+        else:
+            chunk_ranges.append([lower, upper])
+
+    return chunk_ranges
+
+@jit(forceobj=True, fastmath=True)
+def expand_ranges(ranges):
+    length = 0 
+    for bounds in ranges:
+        length += bounds[1] - bounds[0]
+
+    output = np.zeros(length, dtype=np.int64)
+    i = 0
+    for bounds in ranges:
```

Comment:
Yes, I can't imagine that numba will like this - it does not like iterating over 2D arrays. 
 
We should also replace this with a double for loop, so something like:

```python
index = 0

for bound_index in range(len(ranges)):
    lower = bounds[bound_index][0]
    upper = bounds[bound_index][1]
    bound_length = upper - lower

    for value in range(lower, upper):
        output[index] = value
        index += 1
```

Reply:
If you can get away with for bound in bounds that is better but I don't think numba will like that.

Reply:
Can you replace the numpy call with this explicit second loop please?